src/compute-client/src/controller/instance.rs

```diff
@@ -1536,6 +1557,30 @@ where
         ))
     }
 
+    fn handle_copy_to_response(
+        &mut self,
+        sink_id: GlobalId,
+        response: CopyToResponse,
+    ) -> Option<ComputeControllerResponse<T>> {
+        // We might not be tracking this COPY TO because we have already returned a response
+        // from one of the replicas. In that case, we ignore the response.
+        if self.compute.copy_tos.remove(&sink_id) {
+            let result = match response {
+                CopyToResponse::RowCount(count) => Ok(count),
+                CopyToResponse::Error(error) => Err(anyhow::anyhow!(error)),
+                // We do not expect to receive a `Dropped` message here.
+                // A sink is dropped when a response is already received.
+                // And since we can receive only one response, the `Dropped`
+                // never reaches here.
+                // TODO (mouli): should we panic! instead?
+                CopyToResponse::Dropped => Err(anyhow::anyhow!("unexpected error")),
```

Comment:
Wait, I'm not sure that is right! What happens when the controller client (adapter) drops the sink dataflow (via `drop_collections` before the `COPY` is finished? The controller would send an `AllowCompaction([])` command to the replica, which would drop the dataflow possibly before the copy operation completed, so we would receive a `Dropped` response here.

Reply:
So perhaps the right thing to do is to produce an error here saying “query canceled”? I think we already have an AdapterError variant for that.

Reply:
This is handled similarly to the new subscribe now after https://github.com/MaterializeInc/materialize/pull/25179 with a `drop_compute_sinks` call after sending back a `AdapterError::Canceled` to the user.